sdks/python/apache_beam/runners/portability/fn_api_runner/fn_runner.py

```diff
@@ -364,10 +368,36 @@ def run_stages(self,
           bundle_context_manager = execution.BundleContextManager(
               runner_execution_context, stage, self._num_workers)
 
+          assert (
+              runner_execution_context.watermark_manager.get_stage_node(
+                  bundle_context_manager.stage.name
+              ).input_watermark() == timestamp.MAX_TIMESTAMP), (
+              'wrong watermark for %s. Expected %s, but got %s.' % (
+                  runner_execution_context.watermark_manager.get_stage_node(
+                      bundle_context_manager.stage.name),
+                  timestamp.MAX_TIMESTAMP,
+                  runner_execution_context.watermark_manager.get_stage_node(
+                      bundle_context_manager.stage.name
+                  ).input_watermark()
+              )
+          )
+
           stage_results = self._run_stage(
-              runner_execution_context,
-              bundle_context_manager,
+              runner_execution_context, bundle_context_manager)
+
+          assert (
+              runner_execution_context.watermark_manager.get_stage_node(
+                  bundle_context_manager.stage.name
+              ).input_watermark() == timestamp.MAX_TIMESTAMP), (
```

Comment:
initially I'm checking the input_watermark. I'll add a more advanced check in a follow-up change that updates the runner to support per-bundle execution (instead of per-stage)

Reply:
ah thanks for the catch. fixed that.